crispr_read_counts/single_guide_count.py

```diff
@@ -0,0 +1,249 @@
+
+  def _get_sgrna_library_counts_without_stats(self, trim: int, reverse_complementing: bool):
+
+    samfile = self.open_cram_and_get_sample_name()
+    lib_seqs, lib_seq_size = self.get_lib_seq_dict_and_seq_length(reverse_complementing)
+
+    for read in samfile.fetch(until_eof=True):
+      # if the alignment is secondary, supplymentary or  vendor failed, skip it!
+      if read.flag & 2816:
+        continue
+
+      if reverse_complementing:
+        # NOTE: it could be as simple as this: cram_seq = read.get_forward_sequence()[-trim-lib_seq_size:-trim]
+        # but when trim is 0, end index will be "-0", and this will upset python and return an empty string.
+        # so, trim is always added by 1, and the substring became a bit complicated like below.
+        cram_seq = read.get_forward_sequence()
+        cram_seq = cram_seq[-(trim + 1) - (lib_seq_size - 1):-(trim + 1)] + cram_seq[-(trim + 1)]
+      else:
+        cram_seq = read.get_forward_sequence()[trim:trim + lib_seq_size]
+
+      matching_lib_seq = lib_seqs.get(cram_seq)
+      if matching_lib_seq:
+        for grna_id in self.lib[matching_lib_seq]:
+          self.sample_count[grna_id] = self.sample_count.get(grna_id, 0) + 1
+
+  def write_output(self, out_stats: str):
+    if out_stats:
+      self._write_output_with_stats(out_stats)
+    else:
+      self._write_output_without_stats()
+
+  def _write_output_with_stats(self, out_stats: str):
+    zero_count_guides, low_count_guides = 0, 0
+    with open(self.out_count, 'w') as f:
+      if self.plas_name:
+        f.write('\t'.join(['sgRNA', 'gene', f'{self.sample_name}.sample', self.plas_name]) + '\n')
+        for sgrna_seq in (sorted(self.lib.keys())):
+          for sgrna_id in self.lib[sgrna_seq]:
+            count = self.sample_count.get(sgrna_id, 0)
+            if count == 0:
+              zero_count_guides += 1
+            if count < self.LOW_COUNT_GUIDES_THRESHOLD:
+              low_count_guides += 1
+            plasmid_count = self.plasmid.get(sgrna_id, 0)
+            f.write('\t'.join([sgrna_id, self.targeted_genes[sgrna_id], str(count), str(plasmid_count)]) + '\n')
+      else:
+        f.write('\t'.join(['sgRNA', 'gene', f'{self.sample_name}.sample']) + '\n')
+        for sgrna_seq in (sorted(self.lib.keys())):
+          for sgrna_id in self.lib[sgrna_seq]:
+            count = self.sample_count.get(sgrna_id, 0)
+            if count == 0:
+              zero_count_guides += 1
+            if count < self.LOW_COUNT_GUIDES_THRESHOLD:
+              low_count_guides += 1
+            f.write('\t'.join([sgrna_id, self.targeted_genes[sgrna_id], str(count)]) + '\n')
+
+    self.stats['zero_count_guides'] = zero_count_guides
+    self.stats['low_count_guides'] = low_count_guides
+    with open(out_stats, 'w') as out_s:
+      json.dump(self.stats, out_s)
+      out_s.write('\n')
+
+  def _write_output_without_stats(self):
+    with open(self.out_count, 'w') as f:
```

Comment:
Using the `csv` package could simplify the write operations, though I don't know what the performance implications would be (if any).

Reply:
Using CSV does not have obvious performance implications, actually using CSV is marginally quicker.

Reply:
When benchmarking dual_count module, using csv increases run time about 1.5 times when writing out 23 million lines.

Reply:
I wonder if setting a [large] buffer size when opening the output file could improve things for `csv` methods (as well as raw writes). Never tried this, though.

Reply:
Tried 100MB buffer when opening a file to write. It made little difference in performance for csv or row writes...